### ISSUE TITLE
AntiVirus false positives: We figured out how to reduce them

### ISSUE THREAD

**Issue by cculianu** [spesmilo/electrum#5426]
I'm not sure if you guys are concerned with the frequent occurrence of anti-virus false positives for Electrum.  I know on Electron Cash it was causing many of our users to get blocked downloads and other nonsense.

Here is latest electrum 3.3.6 setup .exe's AV score:

https://www.virustotal.com/gui/file/7bc45c53a0179f5889dd68c7f023c6b27e050bf73c84bcd854a6ffe3a83bdf1d/detection

As you can see above.. not great.  We had similar on Electron Cash.

We isolated the problem down to PyInstaller 3.4 (it really only started happening for us when we switched PyInstaller to 3.4 a couple of months ago).

It turns out the `runw.exe` included in PyInstaller (the bootloader) is just on so many anti-virus company lists as "probably bad" (likely due to its frequent use in actual malware) that its hash is nearly perma-banned. :)

So -- we decided to recompile the PyInstaller bootloader as per the PyInstaller docs and just use our own hosted PyInstaller3.4.tar.gz in our Windows build (otherwise identical, just includes only recompiled runw.exe).. **and that did it!**

Now we are down to 1 bogus heuristic detection (down from 15) on a single obscure AV engine:

https://www.virustotal.com/gui/file/88a3defd31d319fd7d65996e2c84bd29ae6d4c6fa6e2e5dcad35792a1193d501/detection

Anyway -- I thought I would give you guys a heads up on this as fake AV detections give the entire crypto space a bad name,

Here is our Electron Cash commit that handles this, for reference:

https://github.com/Electron-Cash/Electron-Cash/compare/9532508a3f46...19eb630a47b6

**Comment by SomberNight** [comment]
Thanks for the info.

@bauerj actually had this idea of compiling our own pyinstaller bootloader before IIRC
We thought Electron Cash already did this though -- weren't there some issues in upstream pyinstaller that you have custom fixes for? I can't recall exactly.

Also, the high AV detection started for Electrum started when we started to use pyinstaller 3 (well before 3.4)
(ref https://github.com/spesmilo/electrum/issues/3198)

I fear if we build our own custom static pyinstaller bootloader and use that every time, it will accumulate a high false positive rate on its own due to all the malware clones of Electrum which would use the same bootloader.
Another point is that we should try to minimise the binary blobs that we use in the build process.

So maybe we could dynamically build the pyinstaller bootloader as part of the Windows/Wine build.

**Comment by cculianu** [comment]
@bauerj actually had this idea of compiling our own pyinstaller bootloader before IIRC
We thought Electron Cash already did this though -- weren't there some issues in upstream pyinstaller that you have custom fixes for? I can't recall exactly.

Yeah on macOS we use a custom PyInstaller (one that incoprorates 2 patches we need to get macOS to not suck).

On Windows we just used the vanilla release one from pypi.

> Also, the high AV detection started for Electrum started when we started to use pyinstaller 3 (well before 3.4) (ref #3198)

Ah.. weird.  We used this PyInstaller from the fix_4442.zip or whatever it's called that was in ThomasV's repo... but a DIFFERENT one than you used (you guys would clone the repo I believe and install that -- we used a .zip form the repo that was pre-packaged), and our false positive count was relatively low then... :/

> I fear if we build our own custom static pyinstaller bootloader and use that every time, it will accumulate a high false positive rate on its own due to all the malware clones of Electrum which would use the same bootloader.
> ..
> So maybe we could dynamically build the pyinstaller bootloader as part of the Windows/Wine build.

That's my concern as well.  And we are working on it!

We're actually working on getting the bootloader to compile every time in the docker build image rather than using our own opaque package.  The problem is right now that the bootloader doesn't like to get compiled with mingw-gcc and really likes to get compiled with MSVC CL.exe.  And installing ms build tools 2015 inside the docker image is not possible because of wine crash bugs when you try to do it. :/

We're working on getting it to play nice with mingw-gcc and if we get it to work -- I'll let you know.  Ideally you recompile it each time -- and maybe even introduce some deterministic nonce in the build -- that would be the best solution albeit complex. :/
